Bottom: 8e858610108557f53e93dd7af44d25857fb890d0
Top:    e06d8cda98db4e12dce26674115aa70458eae4d8
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-07-18 10:10:20 -0700

Refresh of fork

---

diff --git a/src/preload.c b/src/preload.c
index e901d13..7be40f6 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -437,20 +437,23 @@ int listen(int socket, int backlog)
 int accept(int socket, struct sockaddr *addr, socklen_t *addrlen)
 {
 	int fd, index, ret;
+	enum fd_type type;
 
-	if (fd_get(socket, &fd) == fd_rsocket) {
+	type = fd_get(socket, &fd);
+	if (type == fd_rsocket || type == fd_fork) {
 		index = fd_open();
 		if (index < 0)
 			return index;
 
-		ret = raccept(fd, addr, addrlen);
+		ret = (type == fd_rsocket) ? raccept(fd, addr, addrlen) :
+					     real.accept(fd, addr, addrlen);
 		if (ret < 0) {
 			fd_close(index, &fd);
 			return ret;
 		}
 
-		fd_store(index, ret, fd_rsocket);
-		last_accept = index;
+		fd_store(index, ret, type);
+		last_accept = (type == fd_fork) ? index : -1;
 		return index;
 	} else {
 		last_accept = -1;
